File: web/src/api/metrics.ts
// Metrics API calls (docs/features/monitoring.md §1-2; internal/api/metrics.go's
// GET /metrics/live, GET /metrics/history).
import { apiFetch } from "./client";
import type { HistoryPoint, LiveMetric, MetricsHistoryResponse, MetricsLiveResponse } from "./types";

/** GET /metrics/live?refs=a,b,c — current rates for the given entities. An
 * empty `refs` array short-circuits to `[]` without a request: there is
 * nothing meaningful to ask the server for. Refs the sampler hasn't yet
 * seen twice (no rate computable) are simply absent from the result — see
 * internal/metrics.Sampler.Live's doc comment — not an error. */
export function fetchMetricsLive(refs: string[]): Promise<LiveMetric[]> {
  if (refs.length === 0) return Promise.resolve([]);
  const qs = `?refs=${encodeURIComponent(refs.join(","))}`;
  return apiFetch<MetricsLiveResponse>(`/metrics/live${qs}`).then((r) => r.items);
}

/** GET /metrics/history?ref=&fromTs=&toTs= — 24h-ring rate history for one
 * entity. fromTs/toTs are unix seconds; both optional (server-side
 * defaults cover "everything since"/"everything up to"). */
export function fetchMetricsHistory(ref: string, fromTs?: number, toTs?: number): Promise<HistoryPoint[]> {
  const params = new URLSearchParams({ ref });
  if (fromTs !== undefined) params.set("fromTs", String(fromTs));
  if (toTs !== undefined) params.set("toTs", String(toTs));
  return apiFetch<MetricsHistoryResponse>(`/metrics/history?${params.toString()}`).then((r) => r.items);
}

File: internal/api/metrics.go
// SPDX-License-Identifier: Apache-2.0

package api

import (
	"context"
	"net/http"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"

	"github.com/bgovanlu/vnprox/internal/metrics"
)

// MetricsService is the subset of *metrics.Sampler the router needs for
// docs/api.md's Metrics routes (T-601): current rates for a set of
// entities, and 24h-ring history for one entity.
type MetricsService interface {
	Live(refs []string) []metrics.LiveMetric
	History(ctx context.Context, ref string, fromTs, toTs int64) ([]metrics.HistoryPoint, error)
}

// mountMetricsRoutes registers docs/api.md's `GET /metrics/live` and
// `GET /metrics/history` — netRead-gated reads, mounted the same way
// mountFDBRoutes/mountLLDPRoutes are. The `metrics.sample` WS event these
// two REST routes complement is pushed directly by internal/metrics.Sampler
// over the shared /api/ws hub (mountWSRoute above); there is no separate
// mounting step for it here.
func mountMetricsRoutes(r chi.Router, svc MetricsService, auth AuthService) {
	if svc == nil || auth == nil {
		return
	}
	r.Group(func(r chi.Router) {
		r.Use(auth.SessionMiddleware)
		r.Use(auth.RequireCap(capNetRead))
		r.Get("/metrics/live", handleMetricsLive(svc))
		r.Get("/metrics/history", handleMetricsHistory(svc))
	})
}

// handleMetricsLive serves `GET /metrics/live?refs=a,b,c` (docs/api.md's
// Metrics section): current rates for each of the requested entities. Refs
// never yet sampled twice (no rate computable) are simply omitted from the
// result, per Sampler.Live's documented contract — the response is never an
// error for a valid-looking but not-yet-live ref.
func handleMetricsLive(svc MetricsService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		raw := strings.TrimSpace(r.URL.Query().Get("refs"))
		if raw == "" {
			writeJSON(w, http.StatusOK, map[string]any{"items": []metrics.LiveMetric{}})
			return
		}
		parts := strings.Split(raw, ",")
		refs := make([]string, 0, len(parts))
		for _, p := range parts {
			p = strings.TrimSpace(p)
			if p != "" {
				refs = append(refs, p)
			}
		}
		items := svc.Live(refs)
		if items == nil {
			items = []metrics.LiveMetric{}
		}
		writeJSON(w, http.StatusOK, map[string]any{"items": items})
	}
}

// handleMetricsHistory serves `GET /metrics/history?ref=&fromTs=&toTs=`
// (docs/api.md's Metrics section): 24h-ring rate history for one entity.
// Missing/unparsable fromTs/toTs default to "no bound on that side" (0 and
// the max int64 respectively) rather than erroring — a caller that only
// cares about "everything since X" or "everything up to Y" doesn't need to
// know the other bound.
func handleMetricsHistory(svc MetricsService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		ref := strings.TrimSpace(r.URL.Query().Get("ref"))
		if ref == "" {
			writeJSONError(w, http.StatusBadRequest, "validation_failed", "ref is required")
			return
		}
		fromTs := parseTsOrDefault(r.URL.Query().Get("fromTs"), 0)
		toTs := parseTsOrDefault(r.URL.Query().Get("toTs"), 1<<62)

		points, err := svc.History(r.Context(), ref, fromTs, toTs)
		if err != nil {
			writeJSONError(w, http.StatusInternalServerError, "internal", "reading metric history")
			return
		}
		if points == nil {
			points = []metrics.HistoryPoint{}
		}
		writeJSON(w, http.StatusOK, map[string]any{"ref": ref, "items": points})
	}
}

func parseTsOrDefault(raw string, def int64) int64 {
	if raw == "" {
		return def
	}
	v, err := strconv.ParseInt(raw, 10, 64)
	if err != nil {
		return def
	}
	return v
}
